[docs] [dagster-aws] add docs for PipesGlueClient (#22969)
dagster-io/dagster: docs/content/concepts/dagster-pipes/aws-glue.mdx

---

# AWS Glue & Dagster Pipes

This tutorial gives a short overwiew on how to use [Dagster Pipes](/concepts/dagster-pipes) with [AWS Glue](https://aws.amazon.com/glue/).

The [dagster-aws](/\_apidocs/libraries/dagster-aws) integration library provides the [PipesGlueClient](/\_apidocs/libraries/dagster-aws#dagster_aws.pipes.PipesGlueClient) Resource which can be used to launch AWS Glue jobs from Dagster's `asset`s and `op`s. Dagster can receive regular logs and other events like asset materializations from jobs launched with this client. Using it requires minimal code changes on the job side.

## Prerequisites

- A running Dagster deployment (or local instance)
- An AWS Glue job with Python 3.8+ runtime environment

## Usage

1. Provide the `dagster-pipes` module to the AWS Glue job either by installing it in the Glue job environment or by packaging it aling with the job script.

2. Insert the following code into the job script:

```python
from dagster_pipes import open_dagster_pipes, PipesCliArgsParamsLoader, Pipes

with open_dagster_pipes(params_loader=PipesCliArgsParamsLoader()) as pipes:
    pipes.log.info("Hello from AWS Glue job!")

```

> Full example:

Author: I was assuming the user already has a working Glue job. In this case they would only need to insert these lines to get access to Pipes. 

But I agree we can just get to the full example right away

    PipesCliArgsParamsLoader,
    PipesS3ContextLoader,
    PipesS3MessageWriter,
    open_dagster_pipes,
)

client = boto3.client("s3")
context_loader = PipesS3ContextLoader(client)
message_writer = PipesS3MessageWriter(client)
params_loader = PipesCliArgsParamsLoader()

def main():
    with open_dagster_pipes(
        context_loader=context_loader,
        message_writer=message_writer,
        params_loader=params_loader,
    ) as pipes:
        pipes.log.info("Hello from AWS Glue job!")

if __name__ == "__main__":
    main()
```

2. In the Dagster `asset`/`op` code, use the `PipesGlueClient` resource to launch the job:

```python file=/guides/dagster/dagster_pipes/glue/dagster_code.py startafter=start_asset_marker endbefore=end_asset_marker
import os

# dagster_glue_pipes.py
import boto3
from dagster_aws.pipes import PipesGlueClient

from dagster import AssetExecutionContext, asset

@asset
def glue_pipes_asset(
    context: AssetExecutionContext, pipes_glue_client: PipesGlueClient
):
    return pipes_glue_client.run(
        context=context,
        job_name="Example Job",
        arguments={"some_parameter_value": "1"},
    ).get_materialize_result()
```

> And pass the resource to `Definitions`:

Reviewer: ```suggestion
---

## Step 4: Creating Dagster definitions

Next, add the `PipesGlueClient` resource to your project's <PyObject object="Definitions" /> object:
```

- Making this into its own section; title is just a suggestion
- Adding a `PyObject` for `Definitions`

```python file=/guides/dagster/dagster_pipes/glue/dagster_code.py startafter=start_definitions_marker endbefore=end_definitions_marker
from dagster import Definitions  # noqa
from dagster_aws.pipes import PipesGlueContextInjector, PipesS3MessageReader

bucket = os.environ["DAGSTER_GLUE_S3_CONTEXT_BUCKET"]

defs = Definitions(
    assets=[glue_pipes_asset],
    resources={
        "pipes_glue_client": PipesGlueClient(
            context_injector=PipesGlueContextInjector(
                client=boto3.client("s3"),
                bucket=bucket,
            ),
            message_reader=PipesS3MessageReader(
                client=boto3.client("glue"), bucket=bucket
            ),
        )
    },
)
```

> This will launch the AWS Glue job and monitor it's status until it either fails or succeeds. A job fialure will also cause the Dagster Run to fail with an exception.

Reviewer: ```suggestion
This will launch the AWS Glue job and monitor its status until it either fails or succeeds. A job failure will also cause the Dagster run to fail with an exception.
```

Fixing a few typos

- **In AWS**:

  - An existing AWS account
  - An AWS Glue job with a Python 3.8+ runtime environment

---

## Step 1: Provide the dagster-pipes module

Provide the `dagster-pipes` module to the AWS Glue job either by installing it in the Glue job environment or packaging it along with the job script.

---

## Step 2: Add dagster-pipes to the Glue job

Call `open_dagster_pipes` in the Glue job script to create a context that can be used to send messages to Dagster:

```python file=/guides/dagster/dagster_pipes/glue/glue_script.py
import boto3
from dagster_pipes import (
    PipesCliArgsParamsLoader,
    PipesS3ContextLoader,
    PipesS3MessageWriter,
    open_dagster_pipes,
)

client = boto3.client("s3")
context_loader = PipesS3ContextLoader(client)
message_writer = PipesS3MessageWriter(client)
params_loader = PipesCliArgsParamsLoader()

def main():
    with open_dagster_pipes(
        context_loader=context_loader,
        message_writer=message_writer,
        params_loader=params_loader,
    ) as pipes:
        pipes.log.info("Hello from AWS Glue job!")

if __name__ == "__main__":
    main()
```

> ## Step 3: Add the PipesGlueClient to Dagster code

Reviewer: ```suggestion
---

## Step 3: Add the PipesGlueClient to Dagster code
```

My bad for missing this!

In the Dagster asset/op code, use the `PipesGlueClient` resource to launch the job:

```python file=/guides/dagster/dagster_pipes/glue/dagster_code.py startafter=start_asset_marker endbefore=end_asset_marker
import os

# dagster_glue_pipes.py
import boto3
from dagster_aws.pipes import PipesGlueClient

from dagster import AssetExecutionContext, asset

@asset
def glue_pipes_asset(
    context: AssetExecutionContext, pipes_glue_client: PipesGlueClient
):
    return pipes_glue_client.run(
        context=context,
        job_name="Example Job",
        arguments={"some_parameter_value": "1"},
    ).get_materialize_result()
```

This will launch the AWS Glue job and monitor its status until it either fails or succeeds. A job failure will also cause the Dagster run to fail with an exception.

---

> ## Step 4: Creating Dagster definitions

Reviewer: ```suggestion
## Step 4: Create Dagster definitions
```

Sorry if I did this, but lets do this instead so it's the same as the other headings

Next, add the `PipesGlueClient` resource to your project's <PyObject object="Definitions" /> object:

```python file=/guides/dagster/dagster_pipes/glue/dagster_code.py startafter=start_definitions_marker endbefore=end_definitions_marker
from dagster import Definitions  # noqa
from dagster_aws.pipes import PipesGlueContextInjector, PipesS3MessageReader

bucket = os.environ["DAGSTER_GLUE_S3_CONTEXT_BUCKET"]

defs = Definitions(
    assets=[glue_pipes_asset],
    resources={
        "pipes_glue_client": PipesGlueClient(
            client=boto3.client("glue"),
            context_injector=PipesGlueContextInjector(
                client=boto3.client("s3"),
                bucket=bucket,
            ),
            message_reader=PipesS3MessageReader(
                client=boto3.client("s3"), bucket=bucket
            ),
        )
    },
)
```

Dagster will now be able to launch the AWS Glue job from the `glue_pipes_asset` asset.

---

> ## References

Reviewer: ```suggestion
## Related
```

Also sorry for not catching this earlier - we use `Related` for these sections
